contracts/Multi-signature/ApprovalCollector.sol

```diff
@@ -0,0 +1,215 @@
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "./MultiplexingContract.sol";
+
+
+/**
+ * @title DelegateERC20 smart contract
+ * @dev Controls execution of ERC20 token transactions
+ */
+contract ApprovalCollector is Ownable {
+    using SafeMath for uint256;
+
+    MultiplexingContract private multiplexingContract;             //address of controlled contract
+    uint256 constant CALL_LIFETIME = 300000;    //transaction lifetime in seconds
+    uint256 private _transactionNonce = 0;
+
+    enum ContractClass {DEFAULT, CRITICAL}
+
+    struct ContractClassParams {
+        uint neededApprovals;
+        uint neededCriticalApprovals;
+    }
+
+    struct Transaction {
+        mapping (address => bool) approvers;
+        uint256 timestamp;
+        uint arrayIndex;    //index in array of pending transactions
+
+        ContractClassParams classParams;
+
+        bytes transaction;  //function selector + encoded arguments
+        address executor;   //contract, that should perform call
+    }
+
+    bytes32[] _pendingTransactions;
+    mapping (bytes32 => Transaction) private _transactions;
+
+    mapping (address => bool) private _administrators;
+    mapping (address => bool) private _criticalApprovers;
+
+    mapping (address => ContractClass) private _contractClass;
+    mapping (uint => ContractClassParams) private _contractClassParams;
+
+    ////////////////////////////////////////////////////////////////////////////////////
+    event TransactionCreated(bytes32 transactionId);                            //new transaction created
+    event ApprovalReceived(address from, bytes32 transactionId);                //received approval
+    event AllApprovalsReceived(bytes32 transactionId);
+    ////////////////////////////////////////////////////////////////////////////////////
+
+    modifier onlyAdmin()
+    {
+        require ((_administrators[msg.sender] || _criticalApprovers[msg.sender]), "Approver address is not administrator");
+        _;
+    }
+
+    constructor (MultiplexingContract _multiplexingContract) public {
+        _criticalApprovers[msg.sender] = true;
+
+        _contractClassParams[uint(ContractClass.DEFAULT)].neededApprovals = 2;
+        _contractClassParams[uint(ContractClass.DEFAULT)].neededCriticalApprovals = 0;
+
+        _contractClassParams[uint(ContractClass.CRITICAL)].neededApprovals = 3;
+        _contractClassParams[uint(ContractClass.CRITICAL)].neededCriticalApprovals = 1;
+
+        multiplexingContract = _multiplexingContract;
+    }
+
+    /**
+     * @dev Adding transaction for approve
+     *
+     * @param executor Address of contract that should execute transaction
+     * @param transaction encoded function selector and parameters
+     * @return identifier of new transaction that should be approved
+     */
+    function executeTransaction( address executor, bytes memory transaction) public {
```

Comment:
All calls will be performed on the multiplexing contract, so we don't need `executor` parameter.

Reply:
I would extend on this: the multiplexer contract is supposed to take care of the fact different transactions go to different contracts. But on the other hand we want be able to control the `execution class` of transactions. Instead of using the target contract address, we could use the method signature (first 4 bytes of Call-Data)